rest-client/src/main/java/org/jboss/pnc/restclient/AdvancedSCMRepositoryClient.java

```diff
@@ -0,0 +1,131 @@
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.pnc.restclient;
+
+import org.jboss.pnc.client.ClientException;
+import org.jboss.pnc.client.Configuration;
+import org.jboss.pnc.client.RemoteResourceException;
+import org.jboss.pnc.client.SCMRepositoryClient;
+import org.jboss.pnc.dto.notification.RepositoryCreationFailure;
+import org.jboss.pnc.dto.notification.SCMRepositoryCreationSuccess;
+import org.jboss.pnc.dto.requests.CreateAndSyncSCMRequest;
+import org.jboss.pnc.dto.response.RepositoryCreationResponse;
+import org.jboss.pnc.restclient.websocket.VertxWebSocketClient;
+import org.jboss.pnc.restclient.websocket.WebSocketClient;
+
+import java.util.concurrent.CompletableFuture;
+
+import static org.jboss.pnc.restclient.websocket.predicates.SCMRepositoryNotificationPredicates.withFailedTaskId;
+import static org.jboss.pnc.restclient.websocket.predicates.SCMRepositoryNotificationPredicates.withSuccessTaskId;
+
+public class AdvancedSCMRepositoryClient extends SCMRepositoryClient {
+
+    private WebSocketClient webSocketClient = new VertxWebSocketClient();
+
+    public AdvancedSCMRepositoryClient(Configuration configuration) {
+        super(configuration);
+        webSocketClient.connect("ws://" + configuration.getHost() + "/pnc-rest-new/notifications").join();
+    }
+
+    private CompletableFuture<RepositoryCreationFailure> waitForScmCreationFailure(String taskId) {
+
+        return webSocketClient.catchRepositoryCreationFailure(withFailedTaskId(taskId))
+                .whenComplete((x, y) -> webSocketClient.disconnect());
+    }
+
+    private CompletableFuture<SCMRepositoryCreationSuccess> waitForScmCreationSuccess(String taskId) {
+
+        return webSocketClient.catchSCMRepositoryCreationSuccess(withSuccessTaskId(taskId))
+                .whenComplete((x, y) -> webSocketClient.disconnect());
+    }
+
+    /**
+     * If the scm repository is already created and present in the Orch database, a RemoteResourceException is thrown.
+     * It's recommended that you check the presence of the repository before creating it.
+     *
+     * @param request scm to create
+     * @return CompletableFuture of the result, whether it was successful or not
+     * @throws RemoteResourceException Most likely thrown if repository already exists
+     * @throws ClientException if the response from the server is strange: both task id and scm repository is null
+     */
+    public CompletableFuture<SCMCreationResult> createNewAndWait(CreateAndSyncSCMRequest request)
+            throws RemoteResourceException, ClientException {
+
+        RepositoryCreationResponse response = super.createNew(request);
```

Comment:
I'd recommend attaching WS listener before making REST requests. You have a guarantee you'll avoid race condition of the WS message coming earlier than WS listeners being registered.

Reply:
I cannot do that since I need to listen to the task id :(

Reply:
shite, you are right...